File: test/assessment/test_fragmentation_curve.cpp
/***************************************************************************
* Copyright ESIEE Paris (2018)                                             *
*                                                                          *
* Contributor(s) : Benjamin Perret                                         *
*                                                                          *
* Distributed under the terms of the CECILL-B License.                     *
*                                                                          *
* The full license is in the file LICENSE, distributed with this software. *
****************************************************************************/

#include <boost/test/unit_test.hpp>
#include "higra/assessment/fragmentation_curve.hpp"
#include "higra/image/graph_image.hpp"
#include "../test_utils.hpp"

using namespace hg;

BOOST_AUTO_TEST_SUITE(test_fragmentation_curve);

    BOOST_AUTO_TEST_CASE(test_assess_fragmentation_curve_BCE_optimal_cut) {
        tree t(array_1d<index_t>{8, 8, 9, 9, 10, 10, 11, 13, 12, 12, 11, 13, 14, 14, 14});
        array_1d<char> ground_truth{0, 0, 1, 1, 1, 2, 2, 2};

        assesser_optimal_cut_BCE assesser(t, ground_truth, 200);

        BOOST_CHECK(assesser.optimal_number_of_regions() == 3);
        BOOST_CHECK(assesser.number_of_region_ground_truth() == 3);
        BOOST_CHECK(almost_equal(assesser.optimal_score(), (2 + 4.0 / 3 + 2.5) / num_leaves(t)));

        auto res = assesser.fragmentation_curve();
        auto &res_scores = res.scores;
        auto &res_k = res.k;

        array_1d<double> ref_scores{2.75, 4.5, 2 + 4.0 / 3 + 2.5, 2 + 4.0 / 3 + 2, 2 + 4.0 / 3 + 4.0 / 3,
                                    2 + 4.0 / 3 + 4.0 / 3, 4, 3};
        array_1d<size_t> ref_k{1, 2, 3, 4, 5, 6, 7, 8};

        BOOST_CHECK(res_scores == (ref_scores / num_leaves(t)));
        BOOST_CHECK(res_k == ref_k);
    }

    BOOST_AUTO_TEST_CASE(test_assess_optimal_partitions_BCE_optimal_cut) {
        tree t(array_1d<index_t>{8, 8, 9, 9, 10, 10, 11, 13, 12, 12, 11, 13, 14, 14, 14});
        array_1d<char> ground_truth{0, 0, 1, 1, 1, 2, 2, 2};

        assesser_optimal_cut_BCE assesser(t, ground_truth, 200);

        std::vector<array_1d<index_t>> optimal_partitions
                {{0, 0, 0, 0, 0, 0, 0, 0},
                 {0, 0, 0, 0, 1, 1, 1, 1},
                 {0, 0, 1, 1, 2, 2, 2, 2},
                 {0, 0, 1, 1, 2, 2, 2, 3},
                 {0, 0, 1, 1, 2, 2, 3, 4},
                 {0, 0, 1, 1, 2, 3, 4, 5},
                 {0, 0, 1, 2, 3, 4, 5, 6},
                 {0, 1, 2, 3, 4, 5, 6, 7}};

        BOOST_CHECK(is_in_bijection(optimal_partitions[2], assesser.optimal_partition()));

        for (index_t i = 0; i < optimal_partitions.size(); i++) {
            BOOST_CHECK(is_in_bijection(optimal_partitions[i], assesser.optimal_partition(i + 1)));
        }
    }

    BOOST_AUTO_TEST_CASE(test_straight_altitudes_BCE_optimal_cut) {
        tree t(array_1d<index_t>{8, 8, 9, 9, 10, 10, 11, 13, 12, 12, 11, 13, 14, 14, 14});
        array_1d<char> ground_truth{0, 0, 1, 1, 1, 2, 2, 2};

        assesser_optimal_cut_BCE assesser(t, ground_truth, 200);

        auto altitudes = assesser.straightened_altitudes(false);
        array_1d<double> ref_scores{2.75, 4.5, 2 + 4.0 / 3 + 2.5, 2 + 4.0 / 3 + 2, 2 + 4.0 / 3 + 4.0 / 3,
                                    2 + 4.0 / 3 + 4.0 / 3, 4, 3};

        std::vector<array_1d<index_t>> optimal_partitions
                {{0, 0, 0, 0, 0, 0, 0, 0},
                 {0, 0, 0, 0, 1, 1, 1, 1},
                 {0, 0, 1, 1, 2, 2, 2, 2},
                 {0, 0, 1, 1, 2, 2, 2, 3},
                 {0, 0, 1, 1, 2, 2, 3, 4},
                 {0, 0, 1, 1, 2, 2, 3, 4},
                 {0, 0, 1, 2, 3, 4, 5, 6},
                 {0, 1, 2, 3, 4, 5, 6, 7}};

        auto sorted = xt::sort(altitudes);

        for (index_t i = 0; i < optimal_partitions.size(); i++) {
            auto tmp = labelisation_horizontal_cut(t,
                                                   altitudes,
                                                   sorted(root(t) - i));
            BOOST_CHECK(is_in_bijection(optimal_partitions[i], tmp));
        }
    }

BOOST_AUTO_TEST_SUITE_END();

File: python/test/test_assessment/test_fragmentation_curve.py
############################################################################
# Copyright ESIEE Paris (2018)                                             #
#                                                                          #
# Contributor(s) : Benjamin Perret                                         #
#                                                                          #
# Distributed under the terms of the CECILL-B License.                     #
#                                                                          #
# The full license is in the file LICENSE, distributed with this software. #
############################################################################

import unittest
import numpy as np
import higra as hg


class TestFragmentationCurve(unittest.TestCase):

    def test_assess_fragmentation_curve_BCE_optimal_cut(self):
        t = hg.Tree((8, 8, 9, 9, 10, 10, 11, 13, 12, 12, 11, 13, 14, 14, 14))
        ground_truth = np.asarray((0, 0, 1, 1, 1, 2, 2, 2), dtype=np.int32)

        assesser = hg.make_assesser_fragmentation_optimal_cut(t, ground_truth, hg.OptimalCutMeasure.BCE)

        self.assertTrue(assesser.optimal_number_of_regions() == 3)
        self.assertTrue(assesser.number_of_region_ground_truth() == 3)
        self.assertTrue(np.isclose(assesser.optimal_score(), (2 + 4.0 / 3 + 2.5) / t.num_leaves()))

        res_k, res_scores = assesser.fragmentation_curve(False)

        ref_scores = np.asarray((2.75, 4.5, 2 + 4.0 / 3 + 2.5, 2 + 4.0 / 3 + 2, 2 + 4.0 / 3 + 4.0 / 3,
                                 2 + 4.0 / 3 + 4.0 / 3, 4, 3))
        ref_k = np.asarray((1, 2, 3, 4, 5, 6, 7, 8), dtype=np.int32)

        self.assertTrue(np.all(res_scores == (ref_scores / t.num_leaves())))
        self.assertTrue(np.all(res_k == ref_k))

    def test_assess_fragmentation_curve_BCE_optimal_cut_on_rag(self):
        vertex_map = np.asarray((0, 0, 1, 1, 2, 2, 3, 4), dtype=np.int64)
        g = hg.get_4_adjacency_graph((1, 8))
        hg.CptRegionAdjacencyGraph.link(g, None, vertex_map, np.ones((7, )))
        t = hg.Tree((6, 6, 5, 5, 7, 7, 8, 8, 8))
        hg.CptHierarchy.link(t, g)
        ground_truth = np.asarray((0, 0, 1, 1, 1, 2, 2, 2), dtype=np.int32)

        assesser = hg.make_assesser_fragmentation_optimal_cut(t, ground_truth, hg.OptimalCutMeasure.BCE)

        self.assertTrue(assesser.optimal_number_of_regions() == 3)
        self.assertTrue(assesser.number_of_region_ground_truth() == 3)
        self.assertTrue(np.isclose(assesser.optimal_score(), (2 + 4.0 / 3 + 2.5) / t.num_leaves()))

        res_k, res_scores = assesser.fragmentation_curve(False)

        ref_scores = np.asarray((2.75, 4.5, 2 + 4.0 / 3 + 2.5, 2 + 4.0 / 3 + 2, 2 + 4.0 / 3 + 4.0 / 3))
        ref_k = np.asarray((1, 2, 3, 4, 5), dtype=np.int32)

        self.assertTrue(np.all(res_scores == (ref_scores / t.num_leaves())))
        self.assertTrue(np.all(res_k == ref_k))

    def test_assess_optimal_partitions_BCE_optimal_cut(self):
        t = hg.Tree((8, 8, 9, 9, 10, 10, 11, 13, 12, 12, 11, 13, 14, 14, 14))
        ground_truth = np.asarray((0, 0, 1, 1, 1, 2, 2, 2), dtype=np.int32)

        assesser = hg.make_assesser_fragmentation_optimal_cut(t, ground_truth, hg.OptimalCutMeasure.BCE)

        optimal_partitions = [np.asarray((0, 0, 0, 0, 0, 0, 0, 0), dtype=np.int32),
                              np.asarray((0, 0, 0, 0, 1, 1, 1, 1), dtype=np.int32),
                              np.asarray((0, 0, 1, 1, 2, 2, 2, 2), dtype=np.int32),
                              np.asarray((0, 0, 1, 1, 2, 2, 2, 3), dtype=np.int32),
                              np.asarray((0, 0, 1, 1, 2, 2, 3, 4), dtype=np.int32),
                              np.asarray((0, 0, 1, 1, 2, 3, 4, 5), dtype=np.int32),
                              np.asarray((0, 0, 1, 2, 3, 4, 5, 6), dtype=np.int32),
                              np.asarray((0, 1, 2, 3, 4, 5, 6, 7), dtype=np.int32)]

        self.assertTrue(hg.is_in_bijection(optimal_partitions[2], assesser.optimal_partition()))

        for i in range(len(optimal_partitions)):
            self.assertTrue(hg.is_in_bijection(optimal_partitions[i], assesser.optimal_partition(i + 1)))


if __name__ == '__main__':
    unittest.main()
